---
title: "<br> <br> An introduction to the R/Bioconductor Ecosystem"
author: "Joselyn Ch√°vez <br> <br> Twitter: @josschavezf1 <br> josschavezf.netlify.app"
date: "January 28th, 2021"
output:
  xaringan::moon_reader:
    css: ["default", "rladies", "rladies-fonts"]
    lib_dir: libs
    seal: false
---

class: inverse center middle
background-image: url("images/joss.png"), url("images/rladies-bmore.png")
background-size: 140px, 140px
background-position: 90% 90%, 5% 5%


# `r rmarkdown::metadata$title`

## `r rmarkdown::metadata$author`

### `r rmarkdown::metadata$date`


```{r configuracion, include = FALSE}
xaringanExtra::use_clipboard()
```

---
.center[
# Materials are available at:

## http://github.com/josschavezf/intro_bioconductor


# Or you can download them by running the following code in your R session: 
]

```{r, eval = FALSE}
if (!require("usethis")) {
    install.packages("usethis") }

usethis::use_course("josschavezf/intro_bioconductor")
```

---
class: middle center

# Slides are available at

## https://josschavezf.github.io/slides_rladies_bmore/intro-bioconductor.html

---
# What is Bioconductor?

A repository for the analysis of genomic data. 

.center[ 
<img src = "images/sequencing.png" style="width:800px; " />

At date, contains **1974** software packages.
]

---
## Bioconductor have diferent types of packages:

<img src = "images/bioconductorlogo.jpg" style="width:400px; position:absolute; top:500px; left:280px; " />

.pull-left[
* Software (functions)
  - Read sequence files
  - Count
  - Normalize
  - Find differences
  - Plot

* Annotation
  - Databases
  - Organism Sequences
  - Gene identity
]

--

.pull-right[
* Experiment
  - Sequences from a biological experiment.
  - Functions to analyze the data.
  - Lead to a result with biological relevance.

* Workflow
  - Implementation of multiple Bioconductor packages.
  - Show the steps to perform an analysis.

]

---
# How do we access to Bioconductor?

.center[
### https://www.bioconductor.org

<img src = "images/bioconductor.png" style="width:550px;" />
]

---
# How do we access to Bioconductor?

We will use the package **BiocManager**  

Run the following code in your R session to install:

```{r, eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
```

--

**BiocManager** help us to:

* Install any package from Bioconductor, for example the package **Biostrings**:

```{r, eval = FALSE}
BiocManager::install("Biostrings")
```

--

* Keep all your installed Bioconductor packages up to date. 

```{r, eval = FALSE}
BiocManager::install(version = "3.12")
```

<br>
Note: Bioconductor is updated twice a year, around April and October.

---
# The Bioconductor Environment

There are some kind of data, such as sequences and alignments that need be treated in a different way that we use to do with lists or data frames.

Let's see some packages that will help us to deal with genomics data.

* Biostrings
* GenomicRanges
* AnnotationHub
* Gviz

<img src = "images/bioconductorlogo.jpg" style="width:400px; position:absolute; top:500px; left:430px; " />

---
# Manipulation of genomic sequences

The package **Biostrings** have functions to read, write and handling genomic sequences.

## Import data
```{r, eval=FALSE}
eco <- readDNAStringSet("../data/eco.fasta")
```

```{r, echo = FALSE, message=FALSE}
library(Biostrings)
eco <- readDNAStringSet("../data/eco.fasta")
eco
```

---
We can get some attributes of the data, as well as subset some sequences of interest.


* Get the number of sequences
```{r}
length(eco)
```
--
* Get the number of characters on each sequence.
```{r}
nchar(eco)
```
--
* Get the frequency of specific characters 
```{r}
letterFrequency(eco, "GC")
```

---
* Subset sequences 
```{r}
eco[1:2]

subseq(eco,start = 1, end = 10)
```
--
* Translate sequences to amino acids
```{r}
translate(eco$`eco-b0001`)
```

---
## We can store more information from each sequence, (eg.genomic position and chromosome).

.center[
<img src = "images/genesABC.png" style="width:600px; " />
]

---
We need to use a **GenomicRanges** object.

```{r}
library(GenomicRanges)
GRanges(seqnames = c("geneA", "geneB", "geneC"),
        ranges = IRanges(start = c(10, 20, 32),
                         end = c(15,27,42) ),
        strand = c("+", "+", "-") )
```

.center[
<img src = "images/genesABC.png" style="width:400px; " />
]

--
* Note that we are using a special way to define the genomic position (IRanges)

---
# We can add more information by using Rle objects and metadata columns

```{r}
GRanges(seqnames = Rle(c("chr2", "chr2", "chr1", "chr3"), 
                       c(1, 1, 2, 1)),
        ranges = IRanges(1:5, width=10:14,
                         names=head(letters, 5)),
        strand = Rle(strand(c("-", "+", "*")), c( 1,3, 1)),
        score=1:5, GC=seq(1, 0, length=5) )
```

---
# Genomic regions with biological relevance

.center[
<img src = "images/splicing.png" style="width:600px; " />
]
--
```{r, echo=FALSE}
gr <- GRanges(seqnames = "geneM",
              ranges = IRanges(start = c(10, 32),
                               end = c(15,42) ),
              strand = "+",
              exon_id = c(1,2) )
gr
```
---

```{r, echo=FALSE}
gr <- GRanges(seqnames = "geneM",
              ranges = IRanges(start = c(10, 32),
                               end = c(15,42) ),
              strand = "+",
              exon_id = c(1,2) )
gr
```

**range** retrieves the pre-processed gene range (with introns and exons)
```{r}
range(gr)
```

---
# Genomic regions with biological relevance

.center[
<img src = "images/promoter.png" style="width:600px; " />
]

**flank()** retrieves the upstream regulatory region
```{r}
gr <- GRanges(seqnames = "geneO",
              ranges = IRanges(start = c(120, 180),
                               end = c(150,240) ),
              strand = "+",
              exon_id = c(1,2) )
flank(gr, 100)
```

---
class: middle center

# Where can we find genomic data?

--

# AnnotationHub

<img src = "images/journey.jpeg" style="width:350px;" />

---
# AnnotationHub

**AnnotationHub** connect with multiple genomic data providers and give us access to their data.

We use AnnotationHub() function to connect with the database. 

Now, we are ready to explore available data.

```{r, eval = FALSE}
ah = AnnotationHub()

ah$species
```

```{r, echo = FALSE}
library(AnnotationHub)
library(dplyr)
ah = AnnotationHub()
ah$species %>% 
  unique %>% 
  head(10)
```

At date, AnnotationHub contains data from 2643 species.

---
Once you find the desired data, you can download it to your session.

```{r, message = FALSE}
hg_genes <- ah[["AH5036"]]

head(hg_genes)
```

Note that the results is a GRanges object

---
# Annotation packages for specific organisms

**org.Hs.eg.db** Maps Gene identifiers to GenBank Accession Numbers for the Human genome
```{r, eval = FALSE}
x <- org.Hs.eg.db
AnnotationDbi::select(x,
                      keys = "ATRX chromatin remodeler",
                      keytype = "GENENAME",
                      columns = c("ENTREZID", 
                                  "ALIAS", "UNIPROT") )
```

.center[
```{r, echo = FALSE, message=FALSE}
library(org.Hs.eg.db)
x <- org.Hs.eg.db
head(AnnotationDbi::select(x,
                      keys = "ATRX chromatin remodeler",
                      keytype = "GENENAME",
                      columns = c("ENTREZID", "ALIAS", "UNIPROT")
                      ),10 )
```
]

---
# Plotting biological data

The package **Gviz** have functions to make representations of a variety of genomic annotation features, using data from public resources (e.g. ENSEMBL or UCSC) or in-house curated data.

Pros of using Gviz: plotting flexibility.
.center[
```{r, echo=FALSE}
library(Gviz)
data(cpgIslands)
atrack <- AnnotationTrack(cpgIslands, name = "CpG")
gtrack <- GenomeAxisTrack()
gen <- genome(cpgIslands)
chr <- as.character(unique(seqnames(cpgIslands)))
itrack <- IdeogramTrack(genome = gen, chromosome = chr)

data(geneModels)
grtrack <- GeneRegionTrack(geneModels, genome = gen, chromosome = chr,
                           name = "Gene Model",
                           transcriptAnnotation = "symbol",
                           background.title = "brown")
displayPars(grtrack) <- list(background.panel = "#FFFEDB", col = NULL)
plotTracks(list(itrack, gtrack, atrack, grtrack))
```
]

---
# You can submit your own packages!

.center[
<img src = "images/cdsb.png" style="width:400px; " />
<img src = "images/regutoolsteam.png" style="width:500px; " />
http://comunidadbioinfo.github.io
]

--
<img src = "images/regutools1.png" style="width:180px; position:absolute; top:400px; left:140px; " />

--

<img src = "images/regutools2.png" style="width:410px; position:absolute; top:405px; left:350px; " />

.center[
<br><br><br><br><br><br><br><br><br><br>
http://www.bioconductor.org/packages/regutools 
]

---
# How regutools works?

First we need to connect with the database by creating a *regulondb* object:

```{r, echo=FALSE, message=FALSE}
library(regutools)
```

```{r, message=FALSE}
regulondb_conn <- connect_database()

e_coli_regulondb <-
    regulondb(
        database_conn = regulondb_conn,
        organism = "E.coli",
        database_version = "1",
        genome_version = "1"
    )
```

---
# Integration of regutools with the Bioconductor Ecosystem

The function **convert_to_granges()** converts a **regulondb_result** object into a **GRanges** object whenever possible to facilitate the integration with other Bioconductor workflows.

```{r}
res <- get_dataset(
  regulondb = e_coli_regulondb,
  dataset = "GENE",
  attributes = c("posleft", "posright", "strand", "name"),
  filters = list("name" = c("araC","crp","lacI"))
)
convert_to_granges(res)
```


---
Integrating your package results with the Bioconductor Ecosystem facilitates
the use of other packages in downstream steps, for example for plotting.

```{r, eval = FALSE }
grange <- GenomicRanges::GRanges("chr",IRanges::IRanges(5000, 10000))

plot_dna_objects(
    regulondb = e_coli_regulondb,
    grange = grange,
    elements = c("gene", "promoter")
)
```

.center[
```{r,echo=FALSE, fig.height= 5}
e_coli_regulondb <-
    regulondb(
        database_conn = regulondb_conn,
        organism = "chr",
        database_version = "1",
        genome_version = "1"
    )

grange <- GenomicRanges::GRanges("chr",IRanges::IRanges(5000, 10000))

plot_dna_objects(
    regulondb = e_coli_regulondb,
    grange = grange,
    elements = c("gene", "promoter")
)
```
]

---
class: center

<br><br><br><br>

# How do we get help?

--
## browseVignettes() is a good start

browseVignettes("Biostrings")

--

# But...

---
# The Bioconductor community is and infinite source of knowledge

.pull-left[
### Support page of Bioconductor
https://support.bioconductor.org

<img src = "images/support.png" style="width:380px; position:absolute; top:370px; left:50px; " />
]

--

.pull-right[
### Slack channel
<img src = "images/slack.png" style="width:160px; position:absolute; top:190px; left:650px; " />

https://bioc-community.herokuapp.com 

<img src = "images/slack2.png" style="width:380px; position:absolute; top:370px; left:450px; " />
]

---
# Share your contributions with the community!

.pull-left[
<img src = "images/bioc2019logo.png" style="width:100px; position:absolute; top:210px; left:80px; " />
<img src = "images/bioc2019text.png" style="width:220px; position:absolute; top:220px; left:190px; " />
<img src = "images/bioc2019ppt.png" style="width:320px; position:absolute; top:380px; left:80px; " />
]

.pull-right[ 
<img src = "images/bioc2019.png" style="width:290px; position:absolute; top:190px; left:480px; " />
]

--

.pull-right[   
<img src = "images/bioc2021.png" style="width:290px; position:absolute; top:420px; left:480px; " />

<br><br><br><br><br><br><br><br><br><br><br><br>
https://bioc2021.bioconductor.org
]

---
class: inverse middle center

background-image: url("images/joss.png"), url("images/rladies-bmore.png")
background-size: 140px, 140px
background-position: 90% 90%, 5% 5%

# Thanks for your attention!

### Twitter: @josschavezf1

### josschavezf.netlify.app

---
class: inverse middle center

# These slides were made using the [xaringan](https://github.com/yihui/xaringan) package from Yihui Xie and the [rladies theme](https://alison.rbind.io/post/2017-12-18-r-ladies-presentation-ninja/) from Alison Hill.
